### ISSUE TITLE
Search for opportunities to visualize data

### ISSUE THREAD

**Issue by MadeInGer** [sagr4019/ResearchProject#162]
Search for information about training / error visualization. What kind of ways are there to plot data in relation to our problem? Does Keras already implement such opportunities and if not, what libraries exists?

**Comment by MadeInGer** [comment]
I'm afraid I didn't find what I was hoping to find. Most of the articles I found deal with the subject to visualize the model (structure of it) or its accuracy but this doesn't help to understand why self-written programs are not getting classified as expected.

**Comment by jan-christiansen** [comment]
I did not check it in depth, but did you consider this one http://lstm.seas.harvard.edu?

**Comment by MadeInGer** [comment]
No I didn't. I have explicitly searched for Keras implementations. Thanks for the hint. I'll see if it is useful and if we can combine that with Keras.

**Comment by MadeInGer** [comment]
It looks promising, but I do not see anything at the [demo](http://lstm.seas.harvard.edu/client/lstmvis.html?project=00parens&source=states::states1&activation=0.3&cw=30&meta=..&pos=100). 

After installing the tool as of the [manual](https://github.com/HendrikStrobelt/LSTMVis#install) I also do not see the expected dataset that I downloaded and extracted into `data` dir if I open up my browser at http://localhost:8888/client/index.html with both Firefox Quantum and Firefox ESR.

@DanielEggelmann will check if he is able to get it running.

**Comment by jan-christiansen** [comment]
The behaviour is very strange. I managed to load a data set but it did not work when I tried it again. The [children books GRU](http://lstm.seas.harvard.edu/client/lstmvis.html?project=50childbook_gru&source=gates_small::reset1&activation=0.3&cw=30&meta=..&pos=100) worked most of the time for me. I have no idea why it does not always work.

**Comment by DanielEggelmann** [comment]
On a related note:
I tried Tensorboard, which was rather simple to setup: [How do I use the Tensorboard callback of Keras?](https://stackoverflow.com/questions/42112260/how-do-i-use-the-tensorboard-callback-of-keras)

All one has to do, was add a callback to the model: `tensorboard = TensorBoard(log_dir='./logs', histogram_freq=0, write_graph=True, write_images=False)`. Then one can start Tensorboard with `tensorboard --logdir ./logs`.

This will for example plot graphs for accuracy and loss in the training and validation data and update them during training.

**Comment by DanielEggelmann** [comment]
I got the LSTMvis tool running. I used the parens 10k dataset, I can see graphs for two states and the output for one layer, but I don't really know how interpret them.

It is running on my Ubuntu system. I followed the steps from the README:
- Downlod source from github
- Install requirements
- Download & extract dataset
- Run `python2 lstm_server.py -dir data`

**Comment by DanielEggelmann** [comment]
The other dataset seems to work aswell.

**Comment by MadeInGer** [comment]
Great, but funny why it doesn't work for me. I also use Ubuntu with the same dataset and I followed the instructions, but I have to say that I was struggling with installing pip dependencies for a long time. It's running but maybe a dependency is broken without showing an error. I'll try to again reinstall pip and the libraries. 

The histograms in Tensorboard look good, but I'm not sure how they can help us understanding why our own code fragments are not working as test sets for our implementation. LSTMVis looks more specific I think, since it should show single words and offers the possibility to select range of words, which might be helpful if we could interpret them.

Someone wrote a [manual](https://github.com/HendrikStrobelt/LSTMVis/blob/master/docs/keras.md) on how to get LSTMVis running with Keras, which we could use.

**Comment by DanielEggelmann** [comment]
I got it to work somewhat with our model. In the folder `lstm_viz_data` in our team folder you can find a working dataset for LSTMViz. But it is not really helpful, since only one state (neuron?) seems to be recorded, therefore further adjustments are needed.

I tried using the model from the Keras manual, linked in the post above, but it does not learn with our data.

**Comment by DanielEggelmann** [comment]
I was able to fix the problem, now the state of all units is included. Although this bumped the file size up to 200mb, therefore I cannot commit this file to github.

**Comment by MadeInGer** [comment]
I tried to get it running additionally under Windows 7, but I do not see what I should see. 

In the meantime, I did some more research regarding visualizations. Most libraries refer to images. Other libraries/web applications such as http://projector.tensorflow.org/ visualize embeddings to navigate interactively through the data, but I do not see how that can help us currently. I believe that LSTMVis promises the most.

Let's have a teamviewer meeting to explore the visualization. I'll contact you.

**Comment by MadeInGer** [comment]
Together, we have gained an overview of LSTMVis with our model. We saw that the lines in the graphs strongly peak to the top and bottom directly after the declarations in our generated programs. Unfortunately this doesn't give us indication what that means and why this happens.

**Comment by jan-christiansen** [comment]
When I have manually tested your net, I was under the impression that the net might decide based on the variable declarations only. For example, the result abruptly changes for some examples, when changing the order to variable declaration. The peaks might be related to this issue.
